include/nlohmann/detail/iterators/internal_iterator.hpp

```diff
@@ -1,25 +1,372 @@
 
 namespace nlohmann
 {
 namespace detail
 {
+
+template<class ObjectIterator, class ArrayIterator>
+class internal_iterator_base
+{
+  public:
+    enum class iterator_type : std::uint8_t
+    {
+        object,
+        array,
+        primitive,
+        none
+    };
+
+    ObjectIterator& object_iterator()
+    {
+        return *reinterpret_cast<ObjectIterator*>(&m_data);
+    }
+    const ObjectIterator& object_iterator() const
+    {
+        return *reinterpret_cast<const ObjectIterator*>(&m_data);
+    }
+
+    ArrayIterator& array_iterator()
+    {
+        return *reinterpret_cast<ArrayIterator*>(&m_data);
+    }
+    const ArrayIterator& array_iterator() const
+    {
+        return *reinterpret_cast<const ArrayIterator*>(&m_data);
+    }
+
+    primitive_iterator_t& primitive_iterator()
+    {
+        return *reinterpret_cast<primitive_iterator_t*>(&m_data);
+    }
+    const primitive_iterator_t& primitive_iterator() const
+    {
+        return *reinterpret_cast<const primitive_iterator_t*>(&m_data);
+    }
+
+  private:
+    template<class, class, bool>
+    friend class internal_iterator_impl;
+
+    /// Default constructor provided for optimization purposes,
+    /// as it does not initialize the underlying array, should be called internally only.
+    internal_iterator_base() noexcept = default;
+
+    explicit internal_iterator_base(iterator_type it_type) noexcept : m_data()
+    {
+        switch (it_type)
+        {
+            case iterator_type::object:
+                ::new (&object_iterator()) ObjectIterator();
+                break;
+            case iterator_type::array:
+                ::new (&array_iterator()) ArrayIterator();
+                break;
+            case iterator_type::primitive:
+                ::new (&primitive_iterator()) primitive_iterator_t();
+                break;
+            case iterator_type::none:
+                break;
+            default:                 // LCOV_EXCL_LINE
+                JSON_ASSERT(false);  // NOLINT(cert-dcl03-c,hicpp-static-assert,misc-static-assert) LCOV_EXCL_LINE
+        }
+    }
+
+    static inline iterator_type iterator_type_from_value(value_t v_type) noexcept
+    {
+        switch (v_type)
+        {
+            case value_t::object:
+            {
+                return iterator_type::object;
+            }
+
+            case value_t::array:
+            {
+                return iterator_type::array;
+            }
+
+            case value_t::null:
+            case value_t::string:
+            case value_t::boolean:
+            case value_t::number_integer:
+            case value_t::number_unsigned:
+            case value_t::number_float:
+            case value_t::binary:
+            case value_t::discarded:
+            default:
+            {
+                return iterator_type::primitive;
+            }
+        }
+    }
+
+    struct internal_data_t
+    {
+        internal_data_t() noexcept = default;
+
+#ifdef JSON_HAS_CPP_14
```

Comment:
Please document the idea behind this piece of code.

Reply:
And a note about the `std::aligned_storage` deprecation in C++23.